RenderViewTransitionCapture should be able to use its ImageBuffer directly as layer contents.

mattwoodrow opened the pull request:
#### 66eb479d335a74770f939936286a48ae68dca5a7
<pre>
RenderViewTransitionCapture should be able to use its ImageBuffer directly as layer contents.
<a href="https://bugs.webkit.org/show_bug.cgi?id=282424">https://bugs.webkit.org/show_bug.cgi?id=282424</a>
&lt;<a href="https://rdar.apple.com/139050217">rdar://139050217</a>&gt;

Reviewed by Kimmo Kinnunen.

The existing setContentsToImage path is disabled for UI-side compositing, due to
lack of support for a remote NativeImage.  Adds a new option that is supported
with UI-side compositing, setContentsToImageBuffer.

Change RenderLayerBacking to detect RenderViewTransitionCapture as a suitable
image source, and pass the ImageBuffer through to the GraphicsLayer.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setContentsToImageBuffer):
(WebCore::GraphicsLayer::shouldDirectlyCompositeImageBuffer const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setContentsToImage):
(WebCore::GraphicsLayerCA::setContentsToImageBuffer):
(WebCore::GraphicsLayerCA::platformCALayerDelegatesDisplay const):
(WebCore::GraphicsLayerCA::platformCALayerLayerDisplay):
(WebCore::GraphicsLayerCA::updateContentsImage):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::setLayerContentsToImageBuffer):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateContentsRects):
(WebCore::RenderLayerBacking::isDirectlyCompositedImage const):
(WebCore::RenderLayerBacking::updateImageContents):
(WebCore::RenderLayerBacking::contentsBox const):
* Source/WebCore/rendering/RenderViewTransitionCapture.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::ensureBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::shouldDirectlyCompositeImageBuffer const):
(WebKit::GraphicsLayerCARemote::setLayerContentsToImageBuffer):

Canonical link: <a href="https://commits.webkit.org/286139@main">https://commits.webkit.org/286139@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fedbeafa5409927f6183cd498ccc83b7c2aa7559

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54373 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27760 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Running apply-patch; Checked out pull request; compiling") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79373 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26182 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77060 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2158 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58840 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17113 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48996 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64382 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39229 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46296 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21881 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24514 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67439 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22225 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80859 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1364 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67098 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64401 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66398 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16501 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10336 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8494 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2226 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5014 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3175 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2261 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
```diff
@@ -110,7 +110,6 @@
         return;
 
     m_parameters = parameters;
-    clearBackingStore();
```
This is a bit iffy, I can't remember this exactly.
This might be needed in some other cases, like webgl?

mattwoodrow replied:
The general pattern is that we either call `prepareForDisplay` (which drops any buffer handles, and ensures a following `createContextAndPaintContents` to paint contents instead), or we just set a new buffer handle.

Dropping existing buffer handles when we notify of a size/format/opaque-ness change etc shouldn't be necessary. It does mean that paths that use buffer handles need to re-set their handle more often than otherwise needed though, which is why I dropped it here.

This is all complicated (imo) due to the 'backing store' also handling the 'external buffer' path. I think we should pull all the delegate/buffer-handle code out into a separate object, and make the layer own one or the other.

kkinnunen-apple reviewed (approved):
in case the erase stuff is correct

mattwoodrow committed:
RenderViewTransitionCapture should be able to use its ImageBuffer directly as layer contents.
https://bugs.webkit.org/show_bug.cgi?id=282424
<rdar://139050217>

Reviewed by Kimmo Kinnunen.

The existing setContentsToImage path is disabled for UI-side compositing, due to
lack of support for a remote NativeImage.  Adds a new option that is supported
with UI-side compositing, setContentsToImageBuffer.

Change RenderLayerBacking to detect RenderViewTransitionCapture as a suitable
image source, and pass the ImageBuffer through to the GraphicsLayer.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setContentsToImageBuffer):
(WebCore::GraphicsLayer::shouldDirectlyCompositeImageBuffer const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setContentsToImage):
(WebCore::GraphicsLayerCA::setContentsToImageBuffer):
(WebCore::GraphicsLayerCA::platformCALayerDelegatesDisplay const):
(WebCore::GraphicsLayerCA::platformCALayerLayerDisplay):
(WebCore::GraphicsLayerCA::updateContentsImage):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::setLayerContentsToImageBuffer):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateContentsRects):
(WebCore::RenderLayerBacking::isDirectlyCompositedImage const):
(WebCore::RenderLayerBacking::updateImageContents):
(WebCore::RenderLayerBacking::contentsBox const):
* Source/WebCore/rendering/RenderViewTransitionCapture.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::ensureBackingStore):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::shouldDirectlyCompositeImageBuffer const):
(WebKit::GraphicsLayerCARemote::setLayerContentsToImageBuffer):

Canonical link: https://commits.webkit.org/286139@main
